hazelcast/src/main/java/com/hazelcast/nio/tcp/nonblocking/FastSelectionKeys.java

```diff
@@ -0,0 +1,212 @@
+
+    // Using this property the FastSelectionKeys optimization can be disabled. It is disabled by default.
+    private static final boolean ENABLED = Boolean.getBoolean(HAZELCAST_IO_FASTSELECTIONKEYS_ENABLED);
+
+    // the active SelectionKeys is the one where is being added to.
+    SelectionKeys activeKeys = new SelectionKeys();
+    // the passive SelectionKeys is one that is being read using the iterator.
+    SelectionKeys passiveKeys = new SelectionKeys();
+    private final IteratorImpl iterator = new IteratorImpl();
+
+    FastSelectionKeys() {
+    }
+
+    @Override
+    public boolean add(SelectionKey o) {
+        return activeKeys.add(o);
+    }
+
+    @Override
+    public int size() {
+        return activeKeys.size;
+    }
+
+    @Override
+    public Iterator<SelectionKey> iterator() {
+        iterator.init(flip());
+        return iterator;
+    }
+
+    private SelectionKey[] flip() {
+        SelectionKeys tmp = activeKeys;
+        activeKeys = passiveKeys;
+        passiveKeys = tmp;
+
+        activeKeys.size = 0;
+        return passiveKeys.keys;
+    }
+
+    @Override
+    public boolean remove(Object o) {
+        return false;
+    }
+
+    @Override
+    public boolean contains(Object o) {
+        return false;
+    }
+
+    /**
+     * Tries to optimize the provided Selector.
+     *
+     * @param selector
+     * @return an FastSelectionKeySet if the optimization was a success, null otherwise.
+     */
+    public static FastSelectionKeys optimize(Selector selector, ILogger logger) {
+        checkNotNull(selector, "selector");
+        checkNotNull(logger, "logger");
+
+        if (!ENABLED) {
+            logger.finest("FastSelectionKeys optimization has been disabled.");
+            return null;
+        }
+
+        logger.info("FastSelectionKeys is enabled");
+
+        try {
+            FastSelectionKeys selectedKeySet = new FastSelectionKeys();
+
+            Class<?> selectorImplClass = findOptimizableSelectorClass(selector);
+            if (selectorImplClass == null) {
+                return null;
+            }
+
+            Field selectedKeysField = selectorImplClass.getDeclaredField("selectedKeys");
+            selectedKeysField.setAccessible(true);
+
+            Field publicSelectedKeysField = selectorImplClass.getDeclaredField("publicSelectedKeys");
+            publicSelectedKeysField.setAccessible(true);
+
+            selectedKeysField.set(selector, selectedKeySet);
+            publicSelectedKeysField.set(selector, selectedKeySet);
+
+            logger.info("Injected FastSelectionKeys into " + selector.getClass().getName());
+
+            return selectedKeySet;
+        } catch (Throwable t) {
+            // we don't want to print a warning because it could very well be that the target JVM doesn't
+            // support this FastSelectionKeys optimization.
+            logger.warning("Failed to inject FastSelectionKeys into " + selector.getClass().getName(), t);
```

Comment:
this line doesn't match the comment above: You don't want to print a warning, but you are doing it anyway?